crates/turborepo-globwalk/src/lib.rs

```diff
@@ -277,6 +281,79 @@ fn collapse_path(path: &str) -> Option<(Cow<str>, usize)> {
     }
 }
 
+pub fn globwalk(
+    base_path: &AbsoluteSystemPath,
+    include: &[String],
+    exclude: &[String],
+    walk_type: WalkType,
+) -> Result<Vec<Result<AbsoluteSystemPathBuf, WalkError>>, WalkError> {
+    let (base_path_new, include_paths, exclude_paths) =
+        preprocess_paths_and_globs(base_path, include, exclude)?;
+    let inc_patterns = include_paths
+        .iter()
+        .map(|g| Glob::new(g.as_str()).map(|g| g.into_owned()))
+        .collect::<Result<Vec<_>, BuildError>>()?;
+    let ex_patterns = exclude_paths
+        .iter()
+        .map(|g| Glob::new(g.as_str()).map(|g| g.into_owned()))
+        .collect::<Result<Vec<_>, _>>()?;
+
+    let result = inc_patterns
+        .into_iter()
+        .flat_map(|glob| {
+            // Check if the glob specifies an exact filename with no meta characters.
+            if let Some(prefix) = glob.variance().path() {
+                // We expect all of our globs to be absolute paths
+                debug_assert!(glob.has_root());
+                // We're either going to return this path or nothing. Check if it's a directory
+                // and if we want directories
+                match AbsoluteSystemPathBuf::new(prefix)
+                    .map_err(|e| e.into())
+                    .and_then(|path| path.symlink_metadata().map(|md| (path, md)))
+                {
+                    Err(e) => {
+                        // If the file doesn't exist, it's not an error, there's just nothing to
+                        // glob
+                        if e.is_io_error(ErrorKind::NotFound) {
+                            vec![]
+                        } else {
+                            vec![Err(e.into())]
+                        }
+                    }
+                    Ok((path, md)) => {
+                        if walk_type == WalkType::Files && md.is_dir() {
+                            vec![]
+                        } else {
+                            vec![Ok(path)]
+                        }
+                    }
+                }
+            } else {
+                glob.walk(&base_path_new)
+                    .not(ex_patterns.iter().map(|g| g.to_owned()))
+                    // Per docs, only fails if exclusion list is too large, since we're using
+                    // pre-compiled globs
+                    .expect(&format!(
+                        "Failed to compile exclusion globs: {:?}",
+                        ex_patterns
+                    ))
```

Comment:
Avoid the allocation unless necessary
```suggestion
                    .unwrap_or_else(|e| panic!(
                        "Failed to compile exclusion globs: {:?}: {}",
                        ex_patterns,
                        e,
                    ))
```